In [ ]:
import os
import pandas as pd

# Define the data directory path
data_dir = '../data'

In [8]:
# Read CSV files with proper encoding handling
# Try different encodings if utf-8 fails
encodings = ['utf-8', 'latin-1', 'iso-8859-1', 'cp1252']

def read_csv_with_encoding(file_path):
    """Try to read CSV with different encodings"""
    for encoding in encodings:
        try:
            df = pd.read_csv(file_path, encoding=encoding)
            print(f"✓ Loaded {os.path.basename(file_path)} with {encoding} encoding")
            return df
        except UnicodeDecodeError:
            continue
    raise ValueError(f"Could not read {file_path} with any encoding")

# Read all CSV files
dataco_df = read_csv_with_encoding(os.path.join(data_dir, 'DataCoSupplyChainDataset.csv'))
dynamic_logistics_df = read_csv_with_encoding(os.path.join(data_dir, 'dynamic_supply_chain_logistics_dataset.csv'))
supply_chain_df = read_csv_with_encoding(os.path.join(data_dir, 'supply_chain_data.csv'))

# Read Excel file
retail_sales_df = pd.read_excel(os.path.join(data_dir, 'Retail-Supply-Chain-Sales-Dataset.xlsx'))
print(f"✓ Loaded Retail-Supply-Chain-Sales-Dataset.xlsx")

print("\n" + "="*60)
print("All data loaded successfully!")
print("="*60)
print(f"\nDataCoSupplyChainDataset: {dataco_df.shape}")
print(f"Dynamic Supply Chain Logistics: {dynamic_logistics_df.shape}")
print(f"Supply Chain Data: {supply_chain_df.shape}")
print(f"Retail Sales: {retail_sales_df.shape}")

✓ Loaded DataCoSupplyChainDataset.csv with latin-1 encoding
✓ Loaded dynamic_supply_chain_logistics_dataset.csv with utf-8 encoding
✓ Loaded supply_chain_data.csv with utf-8 encoding
✓ Loaded Retail-Supply-Chain-Sales-Dataset.xlsx

All data loaded successfully!

DataCoSupplyChainDataset: (180519, 53)
Dynamic Supply Chain Logistics: (32065, 26)
Supply Chain Data: (100, 24)
Retail Sales: (9994, 23)


In [10]:
dataco_df.head()

Type  Days for shipping (real)  Days for shipment (scheduled)  \
0     DEBIT                         3                              4   
1  TRANSFER                         5                              4   
2      CASH                         4                              4   
3     DEBIT                         3                              4   
4   PAYMENT                         2                              4   

   Benefit per order  Sales per customer   Delivery Status  \
0          91.250000          314.640015  Advance shipping   
1        -249.089996          311.359985     Late delivery   
2        -247.779999          309.720001  Shipping on time   
3          22.860001          304.809998  Advance shipping   
4         134.210007          298.250000  Advance shipping   

   Late_delivery_risk  Category Id   Category Name Customer City  ...  \
0                   0           73  Sporting Goods        Caguas  ...   
1                   1           73  Sporting Goods        Caguas  ...   
2                   0           73  Sporting Goods      San Jose  ...   
3                   0           73  Sporting Goods   Los Angeles  ...   
4                   0           73  Sporting Goods        Caguas  ...   

  Order Zipcode Product Card Id Product Category Id  Product Description  \
0           NaN            1360                  73                  NaN   
1           NaN            1360                  73                  NaN   
2           NaN            1360                  73                  NaN   
3           NaN            1360                  73                  NaN   
4           NaN            1360                  73                  NaN   

                                  Product Image  Product Name Product Price  \
0  http://images.acmesports.sports/Smart+watch   Smart watch         327.75   
1  http://images.acmesports.sports/Smart+watch   Smart watch         327.75   
2  http://images.acmesports.sports/Smart+watch   Smart watch         327.75   
3  http://images.acmesports.sports/Smart+watch   Smart watch         327.75   
4  http://images.acmesports.sports/Smart+watch   Smart watch         327.75   

  Product Status shipping date (DateOrders)   Shipping Mode  
0              0             2/3/2018 22:56  Standard Class  
1              0            1/18/2018 12:27  Standard Class  
2              0            1/17/2018 12:06  Standard Class  
3              0            1/16/2018 11:45  Standard Class  
4              0            1/15/2018 11:24  Standard Class  

[5 rows x 53 columns]

In [12]:
dynamic_logistics_df.head()

timestamp  vehicle_gps_latitude  vehicle_gps_longitude  \
0  2021-01-01 00:00:00             40.375568             -77.014318   
1  2021-01-01 01:00:00             33.507818            -117.036902   
2  2021-01-01 02:00:00             30.020640             -75.269224   
3  2021-01-01 03:00:00             36.649223             -70.190529   
4  2021-01-01 04:00:00             30.001279             -70.012195   

   fuel_consumption_rate  eta_variation_hours  traffic_congestion_level  \
0               5.136512             4.998009                  5.927586   
1               5.101512             0.984929                  1.591992   
2               5.090803             4.972665                  8.787765   
3               8.219558             3.095064                  0.045257   
4               5.000075             3.216077                  8.004851   

   warehouse_inventory_level  loading_unloading_time  \
0                 985.716862                4.951392   
1                 396.700206                1.030379   
2                 832.408935                4.220229   
3                   0.573283                0.530186   
4                 914.925067                3.620890   

   handling_equipment_availability  order_fulfillment_status  ...  \
0                         0.481294                  0.761166  ...   
1                         0.620780                  0.196594  ...   
2                         0.810933                  0.152742  ...   
3                         0.008525                  0.811885  ...   
4                         0.020083                  0.053659  ...   

   iot_temperature  cargo_condition_status  route_risk_level  \
0         0.574400                0.777263          1.182116   
1        -9.753493                0.091839          9.611988   
2        -6.491034                0.253529          6.570431   
3        -0.151276                0.877576          0.548952   
4         2.429448                0.262081          8.861443   

   customs_clearance_time  driver_behavior_score  fatigue_monitoring_score  \
0                0.502006               0.033843                  0.978599   
1                0.966774               0.201725                  0.918586   
2                0.945627               0.264045                  0.394215   
3                4.674035               0.362885                  0.905444   
4                3.445429               0.016957                  0.258702   

   disruption_likelihood_score  delay_probability  risk_classification  \
0                     0.506152           0.885291        Moderate Risk   
1                     0.980784           0.544178            High Risk   
2                     0.998633           0.803322            High Risk   
3                     0.993320           0.025977            High Risk   
4                     0.912433           0.991122            High Risk   

   delivery_time_deviation  
0                 9.110682  
1                 8.175281  
2                 1.283594  
3                 9.304897  
4                 7.752484  

[5 rows x 26 columns]

In [13]:
supply_chain_df.head()

Product type   SKU      Price  Availability  Number of products sold  \
0     haircare  SKU0  69.808006            55                      802   
1     skincare  SKU1  14.843523            95                      736   
2     haircare  SKU2  11.319683            34                        8   
3     skincare  SKU3  61.163343            68                       83   
4     skincare  SKU4   4.805496            26                      871   

   Revenue generated Customer demographics  Stock levels  Lead times  \
0        8661.996792            Non-binary            58           7   
1        7460.900065                Female            53          30   
2        9577.749626               Unknown             1          10   
3        7766.836426            Non-binary            23          13   
4        2686.505152            Non-binary             5           3   

   Order quantities  ...  Location Lead time  Production volumes  \
0                96  ...    Mumbai        29                 215   
1                37  ...    Mumbai        23                 517   
2                88  ...    Mumbai        12                 971   
3                59  ...   Kolkata        24                 937   
4                56  ...     Delhi         5                 414   

  Manufacturing lead time Manufacturing costs  Inspection results  \
0                      29           46.279879             Pending   
1                      30           33.616769             Pending   
2                      27           30.688019             Pending   
3                      18           35.624741                Fail   
4                       3           92.065161                Fail   

   Defect rates  Transportation modes   Routes       Costs  
0      0.226410                  Road  Route B  187.752075  
1      4.854068                  Road  Route B  503.065579  
2      4.580593                   Air  Route C  141.920282  
3      4.746649                  Rail  Route A  254.776159  
4      3.145580                   Air  Route A  923.440632  

[5 rows x 24 columns]

In [14]:
retail_sales_df.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-08-11 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-08-11 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-12-06 2016-12-06    Second Class    DV-13045   
3       4  US-2015-108966 2015-11-10 2015-11-10  Standard Class    SO-20335   
4       5  US-2015-108966 2015-11-10 2015-11-10  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

  Retail Sales People       Product ID         Category Sub-Category  \
0   Cassandra Brandow  FUR-BO-10001798        Furniture    Bookcases   
1   Cassandra Brandow  FUR-CH-10000454        Furniture       Chairs   
2       Anna Andreadi  OFF-LA-10000240  Office Supplies       Labels   
3   Cassandra Brandow  FUR-TA-10000577        Furniture       Tables   
4   Cassandra Brandow  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name Returned     Sales  \
0                  Bush Somerset Collection Bookcase      Not  261.9600   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...      Not  731.9400   
2  Self-Adhesive Address Labels for Typewriters b...      Not   14.6200   
3      Bretford CR4500 Series Slim Rectangular Table      Not  957.5775   
4                     Eldon Fold 'N Roll Cart System      Not   22.3680   

  Quantity Discount    Profit  
0        2     0.00   41.9136  
1        3     0.00  219.5820  
2        2     0.00    6.8714  
3        5     0.45 -383.0310  
4        2     0.20    2.5164  

[5 rows x 23 columns]

# Dataset Analysis for Multi-Agent Supply Chain Intelligence Platform

## Overview of Available Datasets

Based on the loaded data, we have 4 datasets from Kaggle:
1. **DataCo Supply Chain Dataset** (180,519 rows × 53 columns)
2. **Dynamic Logistics Dataset** (32,065 rows × 26 columns)
3. **Supply Chain Data** (100 rows × 24 columns)
4. **Retail Supply Chain Sales Dataset** (9,994 rows × 23 columns)

In [15]:
# Detailed analysis of each dataset
print("="*80)
print("1. DATACO SUPPLY CHAIN DATASET")
print("="*80)
print(f"\nShape: {dataco_df.shape}")
print(f"\nColumns ({len(dataco_df.columns)}):")
print(dataco_df.columns.tolist())
print(f"\nData Types:")
print(dataco_df.dtypes)
print(f"\nSample Data:")
print(dataco_df.head(3))

1. DATACO SUPPLY CHAIN DATASET

Shape: (180519, 53)

Columns (53):
['Type', 'Days for shipping (real)', 'Days for shipment (scheduled)', 'Benefit per order', 'Sales per customer', 'Delivery Status', 'Late_delivery_risk', 'Category Id', 'Category Name', 'Customer City', 'Customer Country', 'Customer Email', 'Customer Fname', 'Customer Id', 'Customer Lname', 'Customer Password', 'Customer Segment', 'Customer State', 'Customer Street', 'Customer Zipcode', 'Department Id', 'Department Name', 'Latitude', 'Longitude', 'Market', 'Order City', 'Order Country', 'Order Customer Id', 'order date (DateOrders)', 'Order Id', 'Order Item Cardprod Id', 'Order Item Discount', 'Order Item Discount Rate', 'Order Item Id', 'Order Item Product Price', 'Order Item Profit Ratio', 'Order Item Quantity', 'Sales', 'Order Item Total', 'Order Profit Per Order', 'Order Region', 'Order State', 'Order Status', 'Order Zipcode', 'Product Card Id', 'Product Category Id', 'Product Description', 'Product Image', 'Product

## Agent Capability Assessment

Let's assess whether these datasets support the 5 agent types you want to build:
- **Data Analyst Agent**: Trends, outliers, anomalies, SQL queries, auto-charts
- **Supply Chain Risk Agent**: Supplier delays, stockout prediction, demand spikes, reorder quantities
- **Finance Insight Agent**: P&L, expense patterns, cashflow forecasting
- **Meeting/Report Agent**: Weekly reports, CxO summaries, recommended actions
- **Email/Workflow Agent**: Stakeholder alerts, follow-up tasks, meeting agendas

In [16]:
# Analyze key features for each agent type

print("🔍 DATASET CAPABILITY ANALYSIS FOR MULTI-AGENT SYSTEM\n")
print("="*80)

# 1. DataCo Supply Chain Dataset Analysis
print("\n📦 Dataset 1: DataCo Supply Chain (Main Dataset)")
print("-"*80)
dataco_features = {
    'Sales & Orders': [col for col in dataco_df.columns if any(x in col.lower() for x in ['sales', 'order', 'customer', 'product'])],
    'Shipping & Logistics': [col for col in dataco_df.columns if any(x in col.lower() for x in ['ship', 'delivery', 'days'])],
    'Financial': [col for col in dataco_df.columns if any(x in col.lower() for x in ['benefit', 'price', 'payment'])],
    'Risk Indicators': [col for col in dataco_df.columns if any(x in col.lower() for x in ['late', 'risk', 'status'])]
}

for category, cols in dataco_features.items():
    print(f"\n  {category}: {len(cols)} columns")
    print(f"    {cols[:5]}")  # Show first 5

# Check for time series capability
print(f"\n  ⏰ Time-series columns:")
time_cols = [col for col in dataco_df.columns if 'date' in col.lower() or 'time' in col.lower()]
print(f"    {time_cols}")

# Check data coverage
print(f"\n  📊 Data Coverage:")
if 'order date (DateOrders)' in dataco_df.columns:
    dataco_df['order_date_parsed'] = pd.to_datetime(dataco_df['order date (DateOrders)'], errors='coerce')
    print(f"    Date Range: {dataco_df['order_date_parsed'].min()} to {dataco_df['order_date_parsed'].max()}")
    print(f"    Duration: {(dataco_df['order_date_parsed'].max() - dataco_df['order_date_parsed'].min()).days} days")

🔍 DATASET CAPABILITY ANALYSIS FOR MULTI-AGENT SYSTEM


📦 Dataset 1: DataCo Supply Chain (Main Dataset)
--------------------------------------------------------------------------------

  Sales & Orders: 40 columns
    ['Benefit per order', 'Sales per customer', 'Customer City', 'Customer Country', 'Customer Email']

  Shipping & Logistics: 6 columns
    ['Days for shipping (real)', 'Days for shipment (scheduled)', 'Delivery Status', 'Late_delivery_risk', 'shipping date (DateOrders)']

  Financial: 3 columns
    ['Benefit per order', 'Order Item Product Price', 'Product Price']

  Risk Indicators: 4 columns
    ['Delivery Status', 'Late_delivery_risk', 'Order Status', 'Product Status']

  ⏰ Time-series columns:
    ['order date (DateOrders)', 'shipping date (DateOrders)']

  📊 Data Coverage:
    Date Range: 2015-01-01 00:00:00 to 2018-01-31 23:38:00
    Duration: 1126 days


In [17]:
# 2. Dynamic Logistics Dataset Analysis
print("\n\n🚚 Dataset 2: Dynamic Logistics Dataset")
print("-"*80)
print(f"  Rows: {dynamic_logistics_df.shape[0]:,}")
print(f"  Columns: {dynamic_logistics_df.shape[1]}")

logistics_features = {
    'Vehicle & GPS': [col for col in dynamic_logistics_df.columns if any(x in col.lower() for x in ['vehicle', 'gps', 'fuel'])],
    'Warehouse & Inventory': [col for col in dynamic_logistics_df.columns if any(x in col.lower() for x in ['warehouse', 'inventory', 'loading'])],
    'Risk & Performance': [col for col in dynamic_logistics_df.columns if any(x in col.lower() for x in ['risk', 'delay', 'disruption'])],
    'Real-time Monitoring': [col for col in dynamic_logistics_df.columns if any(x in col.lower() for x in ['iot', 'temperature', 'fatigue', 'behavior'])]
}

for category, cols in logistics_features.items():
    print(f"\n  {category}: {len(cols)} columns")
    print(f"    {cols}")

# Time range
if 'timestamp' in dynamic_logistics_df.columns:
    dynamic_logistics_df['timestamp_parsed'] = pd.to_datetime(dynamic_logistics_df['timestamp'], errors='coerce')
    print(f"\n  ⏰ Time Range: {dynamic_logistics_df['timestamp_parsed'].min()} to {dynamic_logistics_df['timestamp_parsed'].max()}")
    print(f"  📊 Duration: {(dynamic_logistics_df['timestamp_parsed'].max() - dynamic_logistics_df['timestamp_parsed'].min()).days} days")



🚚 Dataset 2: Dynamic Logistics Dataset
--------------------------------------------------------------------------------
  Rows: 32,065
  Columns: 26

  Vehicle & GPS: 3 columns
    ['vehicle_gps_latitude', 'vehicle_gps_longitude', 'fuel_consumption_rate']

  Warehouse & Inventory: 2 columns
    ['warehouse_inventory_level', 'loading_unloading_time']

  Risk & Performance: 4 columns
    ['route_risk_level', 'disruption_likelihood_score', 'delay_probability', 'risk_classification']

  Real-time Monitoring: 3 columns
    ['iot_temperature', 'driver_behavior_score', 'fatigue_monitoring_score']

  ⏰ Time Range: 2021-01-01 00:00:00 to 2024-08-29 00:00:00
  📊 Duration: 1336 days


In [18]:
# 3. Retail Supply Chain Sales Dataset
print("\n\n🛒 Dataset 3: Retail Supply Chain Sales")
print("-"*80)
print(f"  Rows: {retail_sales_df.shape[0]:,}")
print(f"  Columns: {retail_sales_df.shape[1]}")
print(f"\n  Key Features:")
print(f"    - Orders, Customers, Products")
print(f"    - Sales, Profit, Discount")
print(f"    - Categories, Sub-categories")
print(f"    - Returns tracking")
print(f"    - Geographical data (Country, City, Region)")

if 'Order Date' in retail_sales_df.columns:
    retail_sales_df['Order Date'] = pd.to_datetime(retail_sales_df['Order Date'], errors='coerce')
    print(f"\n  ⏰ Date Range: {retail_sales_df['Order Date'].min()} to {retail_sales_df['Order Date'].max()}")
    print(f"  📊 Duration: {(retail_sales_df['Order Date'].max() - retail_sales_df['Order Date'].min()).days} days")
    
# Financial summary
if 'Sales' in retail_sales_df.columns and 'Profit' in retail_sales_df.columns:
    print(f"\n  💰 Financial Overview:")
    print(f"    Total Sales: ${retail_sales_df['Sales'].sum():,.2f}")
    print(f"    Total Profit: ${retail_sales_df['Profit'].sum():,.2f}")
    print(f"    Profit Margin: {(retail_sales_df['Profit'].sum() / retail_sales_df['Sales'].sum() * 100):.2f}%")



🛒 Dataset 3: Retail Supply Chain Sales
--------------------------------------------------------------------------------
  Rows: 9,994
  Columns: 23

  Key Features:
    - Orders, Customers, Products
    - Sales, Profit, Discount
    - Categories, Sub-categories
    - Returns tracking
    - Geographical data (Country, City, Region)

  ⏰ Date Range: 2014-01-02 00:00:00 to 2017-12-30 00:00:00
  📊 Duration: 1458 days

  💰 Financial Overview:
    Total Sales: $2,297,200.86
    Total Profit: $286,397.02
    Profit Margin: 12.47%


In [19]:
# 4. Supply Chain Data (Smaller dataset)
print("\n\n📊 Dataset 4: Supply Chain Data (Small)")
print("-"*80)
print(f"  Rows: {supply_chain_df.shape[0]:,}")
print(f"  Columns: {supply_chain_df.shape[1]}")
print(f"\n  Key Features:")
print(f"    Columns: {supply_chain_df.columns.tolist()}")
print(f"\n  Note: This is a small dataset (100 rows) - good for testing/validation")



📊 Dataset 4: Supply Chain Data (Small)
--------------------------------------------------------------------------------
  Rows: 100
  Columns: 24

  Key Features:
    Columns: ['Product type', 'SKU', 'Price', 'Availability', 'Number of products sold', 'Revenue generated', 'Customer demographics', 'Stock levels', 'Lead times', 'Order quantities', 'Shipping times', 'Shipping carriers', 'Shipping costs', 'Supplier name', 'Location', 'Lead time', 'Production volumes', 'Manufacturing lead time', 'Manufacturing costs', 'Inspection results', 'Defect rates', 'Transportation modes', 'Routes', 'Costs']

  Note: This is a small dataset (100 rows) - good for testing/validation


## Assessment: Can We Build the Multi-Agent System?

In [20]:
print("\n" + "="*80)
print("✅ FEASIBILITY ASSESSMENT FOR EACH AGENT")
print("="*80)

agents = {
    "1. Data Analyst Agent": {
        "Required Capabilities": [
            "Identify trends, outliers, anomalies",
            "Run SQL queries on demand",
            "Auto-generate charts"
        ],
        "Data Support": "✅ EXCELLENT",
        "Details": [
            "✅ 180K+ rows of transactional data (DataCo)",
            "✅ Time-series data spanning multiple years",
            "✅ Multiple metrics: sales, orders, shipping, customers",
            "✅ Can detect anomalies in delivery times, sales patterns",
            "✅ Rich categorical data for segmentation"
        ]
    },
    
    "2. Supply Chain Risk Agent": {
        "Required Capabilities": [
            "Detect supplier delays",
            "Predict stockouts",
            "Flag demand spikes",
            "Suggest reorder quantities"
        ],
        "Data Support": "✅ VERY GOOD",
        "Details": [
            "✅ Late delivery risk column (DataCo)",
            "✅ Delivery status tracking",
            "✅ Days for shipping (actual vs scheduled)",
            "✅ Warehouse inventory levels (Dynamic Logistics)",
            "✅ Stock levels and lead times (Supply Chain Data)",
            "⚠️  Limited supplier-specific data",
            "⚠️  Would benefit from more procurement data"
        ]
    },
    
    "3. Finance Insight Agent": {
        "Required Capabilities": [
            "P&L summarization",
            "Expense pattern analysis",
            "Cashflow forecasting"
        ],
        "Data Support": "✅ GOOD",
        "Details": [
            "✅ Sales, Revenue, Profit data (Retail Sales)",
            "✅ Benefit per order (DataCo)",
            "✅ Product pricing, costs",
            "✅ Payment types (DEBIT, CASH, TRANSFER)",
            "⚠️  Limited expense breakdown data",
            "⚠️  No direct cashflow/AR/AP data",
            "💡 Can calculate margins, profitability trends"
        ]
    },
    
    "4. Meeting/Report Agent": {
        "Required Capabilities": [
            "Weekly supply chain reports",
            "CxO-level summaries",
            "Recommend 3 actions"
        ],
        "Data Support": "✅ EXCELLENT",
        "Details": [
            "✅ Can aggregate data from all agents",
            "✅ Rich data for KPI dashboards",
            "✅ Time-series for trend reporting",
            "✅ Multiple business dimensions available"
        ]
    },
    
    "5. Email/Workflow Agent": {
        "Required Capabilities": [
            "Send stakeholder alerts",
            "Create follow-up tasks",
            "Generate meeting agendas"
        ],
        "Data Support": "✅ EXCELLENT",
        "Details": [
            "✅ Can trigger on risk events (late deliveries)",
            "✅ Can trigger on inventory thresholds",
            "✅ Can create alerts from other agents' insights",
            "✅ Customer/order data for context"
        ]
    }
}

for agent_name, info in agents.items():
    print(f"\n{agent_name}")
    print("-" * 80)
    print(f"Status: {info['Data Support']}")
    print(f"\nRequired Capabilities:")
    for cap in info['Required Capabilities']:
        print(f"  • {cap}")
    print(f"\nData Support Details:")
    for detail in info['Details']:
        print(f"  {detail}")

print("\n" + "="*80)
print("🎯 OVERALL VERDICT")
print("="*80)
print("\n✅ YES - You CAN build this multi-agent system with the available data!")
print("\nStrengths:")
print("  • Large transaction dataset (180K+ rows)")
print("  • Time-series data for trends and forecasting")
print("  • Logistics and delivery tracking")
print("  • Financial metrics (sales, profit, costs)")
print("  • Risk indicators (late delivery, stockouts)")
print("  • Real-time monitoring data (IoT, GPS)")
print("\nGaps & Workarounds:")
print("  ⚠️  Limited supplier-specific data → Focus on delivery performance")
print("  ⚠️  No direct AR/AP data → Use sales/profit as proxy for cashflow")
print("  ⚠️  Limited procurement data → Focus on inventory and reorder signals")
print("\nRecommendation:")
print("  🚀 START BUILDING! The data supports 90% of your agent requirements.")
print("  💡 Use vector embeddings for the knowledge base")
print("  💡 Implement time-series models for forecasting")
print("  💡 Use LangGraph or CrewAI for agent orchestration")


✅ FEASIBILITY ASSESSMENT FOR EACH AGENT

1. Data Analyst Agent
--------------------------------------------------------------------------------
Status: ✅ EXCELLENT

Required Capabilities:
  • Identify trends, outliers, anomalies
  • Run SQL queries on demand
  • Auto-generate charts

Data Support Details:
  ✅ 180K+ rows of transactional data (DataCo)
  ✅ Time-series data spanning multiple years
  ✅ Multiple metrics: sales, orders, shipping, customers
  ✅ Can detect anomalies in delivery times, sales patterns
  ✅ Rich categorical data for segmentation

2. Supply Chain Risk Agent
--------------------------------------------------------------------------------
Status: ✅ VERY GOOD

Required Capabilities:
  • Detect supplier delays
  • Predict stockouts
  • Flag demand spikes
  • Suggest reorder quantities

Data Support Details:
  ✅ Late delivery risk column (DataCo)
  ✅ Delivery status tracking
  ✅ Days for shipping (actual vs scheduled)
  ✅ Warehouse inventory levels (Dynamic Logistics)


## Key Insights About Your Data

In [21]:
print("="*80)
print("📋 SUMMARY OF DATASETS")
print("="*80)

summary_data = {
    'Dataset': ['DataCo Supply Chain', 'Dynamic Logistics', 'Retail Sales', 'Supply Chain Data'],
    'Rows': [f"{dataco_df.shape[0]:,}", f"{dynamic_logistics_df.shape[0]:,}", 
             f"{retail_sales_df.shape[0]:,}", f"{supply_chain_df.shape[0]:,}"],
    'Columns': [dataco_df.shape[1], dynamic_logistics_df.shape[1], 
                retail_sales_df.shape[1], supply_chain_df.shape[1]],
    'Primary Use': [
        'Sales transactions, orders, shipping',
        'Real-time logistics, GPS, IoT monitoring',
        'Retail operations, profitability',
        'Manufacturing, procurement (small test set)'
    ],
    'Best For Agent': [
        'Data Analyst, Finance, Risk',
        'Risk, Real-time monitoring',
        'Finance, Data Analyst',
        'Testing/Validation'
    ]
}

summary_df = pd.DataFrame(summary_data)
print("\n")
display(summary_df)

print("\n" + "="*80)
print("🔑 KEY DATA POINTS AVAILABLE")
print("="*80)
print("\n📊 For Data Analyst Agent:")
print("  • Order volumes, sales trends over time")
print("  • Customer segmentation (city, state, country)")
print("  • Product categories and performance")
print("  • Shipping modes and delivery patterns")

print("\n⚠️  For Supply Chain Risk Agent:")
print("  • Late delivery risk scores")
print("  • Actual vs scheduled shipping days")
print("  • Warehouse inventory levels")
print("  • Disruption likelihood scores")
print("  • Route risk levels")
print("  • Stock levels and lead times")

print("\n💰 For Finance Insight Agent:")
print("  • Sales revenue by product/customer/region")
print("  • Profit and profit margins")
print("  • Discount patterns")
print("  • Benefit per order")
print("  • Product costs and pricing")

print("\n📈 For Reporting Agent:")
print("  • Time-series data for weekly/monthly reports")
print("  • KPIs: delivery performance, sales, profit")
print("  • Regional performance metrics")
print("  • Return rates")

print("\n🔔 For Workflow/Alert Agent:")
print("  • Late delivery triggers")
print("  • Low inventory alerts")
print("  • High-risk shipment flags")
print("  • Demand spike detection")

📋 SUMMARY OF DATASETS




Dataset     Rows  Columns  \
0  DataCo Supply Chain  180,519       54   
1    Dynamic Logistics   32,065       27   
2         Retail Sales    9,994       23   
3    Supply Chain Data      100       24   

                                   Primary Use               Best For Agent  
0         Sales transactions, orders, shipping  Data Analyst, Finance, Risk  
1     Real-time logistics, GPS, IoT monitoring   Risk, Real-time monitoring  
2             Retail operations, profitability        Finance, Data Analyst  
3  Manufacturing, procurement (small test set)           Testing/Validation


🔑 KEY DATA POINTS AVAILABLE

📊 For Data Analyst Agent:
  • Order volumes, sales trends over time
  • Customer segmentation (city, state, country)
  • Product categories and performance
  • Shipping modes and delivery patterns

⚠️  For Supply Chain Risk Agent:
  • Late delivery risk scores
  • Actual vs scheduled shipping days
  • Warehouse inventory levels
  • Disruption likelihood scores
  • Route risk levels
  • Stock levels and lead times

💰 For Finance Insight Agent:
  • Sales revenue by product/customer/region
  • Profit and profit margins
  • Discount patterns
  • Benefit per order
  • Product costs and pricing

📈 For Reporting Agent:
  • Time-series data for weekly/monthly reports
  • KPIs: delivery performance, sales, profit
  • Regional performance metrics
  • Return rates

🔔 For Workflow/Alert Agent:
  • Late delivery triggers
  • Low inventory alerts
  • High-risk shipment flags
  • Demand spike detection


## Next Steps: Building Your Multi-Agent System

Based on this analysis, here's what you should do next:

### 1. Data Preparation
- Clean and merge datasets where appropriate
- Create unified time-series structures
- Build vector embeddings for knowledge base (use ChromaDB or Pinecone)
- Set up SQL database for agent queries

### 2. Agent Framework Selection
Choose one of these:
- **LangGraph** (Recommended) - Great for complex agentic workflows
- **CrewAI** - Easy to get started, good for role-based agents
- **AutoGen** - Microsoft's framework, good for multi-agent collaboration
- **Custom with LangChain** - Maximum flexibility

### 3. Implementation Priority
1. Start with **Data Analyst Agent** (foundation for others)
2. Build **Supply Chain Risk Agent** (high value)
3. Add **Finance Insight Agent** 
4. Create **Meeting/Report Agent** (aggregates insights)
5. Finally, **Email/Workflow Agent** (orchestration layer)

### 4. Tech Stack Recommendation
- **LLM**: OpenAI GPT-4 or Claude (for agent reasoning)
- **Vector DB**: ChromaDB or Pinecone (for knowledge base)
- **Database**: PostgreSQL or DuckDB (for SQL queries)
- **Visualization**: Plotly or Matplotlib (auto-charts)
- **Orchestration**: LangGraph or CrewAI
- **Monitoring**: LangSmith or custom logging

# Opik Testing

In [22]:
from opik import track
import opik
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv(override=True)

# Initialize Opik
client = opik.Opik()

@track(project_name="omnisupply-data-analyst")
def test_tracking():
    """Test Opik tracking"""
    return {"status": "success", "message": "Opik is working!"}

if __name__ == "__main__":
    result = test_tracking()
    print(result)
    print("\n✅ Check your Comet workspace for traces!")
    print("🔗 https://www.comet.com/omnisupply/projects/omnisupply-data-analyst")

OPIK: Started logging traces to the "omnisupply-data-analyst" project at https://www.comet.com/opik/api/v1/session/redirect/projects/?trace_id=019ae064-68b5-7844-ab6d-2fafea977deb&path=aHR0cHM6Ly93d3cuY29tZXQuY29tL29waWsvYXBpLw==.


{'status': 'success', 'message': 'Opik is working!'}

✅ Check your Comet workspace for traces!
🔗 https://www.comet.com/omnisupply/projects/omnisupply-data-analyst
